### DISCUSSION TITLE
collagraph QScrollArea

### DISCUSSION THREAD

**Discussion by jamesdbrock** [fork-tongue/collagraph#110] [General]
I want a `QScrollArea` which contains a `QBoxLayout`. 

```python
        return cg.h(
            "QScrollArea",
            {
                "vertical-scroll-bar-policy": Qt.ScrollBarAlwaysOn,
            },
            cg.h(
                "Widget",
                {
                    "layout": {
                        "type": "Box",
                        "direction": "TopToBottom",
                        "alignment": Qt.AlignTop,
                    },
                },
```

I'm having the same problem as this guy.

https://stackoverflow.com/questions/12781407/how-do-i-resize-the-contents-of-a-qscrollarea-as-more-widgets-are-placed-inside?rq=3

The top voted answer to use `QScrollArea::setWidget` agrees with this blog post. https://www.pythonguis.com/tutorials/pyside-qscrollarea/

But how can I set the `widget` property?

```python
        return cg.h(
            "QScrollArea",
            {
                "vertical-scroll-bar-policy": Qt.ScrollBarAlwaysOn,
                "widget": cg.h(
                    "Widget",
                    {
```

I cannot set the `widget` property this way because `TypeError: 'PySide6.QtWidgets.QScrollArea.setWidget' called with wrong argument types:  PySide6.QtWidgets.QScrollArea.setWidget(VNode)`

**Comment by berendkleinhaneveld** [comment]
This would require some extra code, like here: https://github.com/fork-tongue/collagraph/blob/master/collagraph/renderers/pyside/objects/qobject.py to implement how to add/set a widget to a scroll area.

**Reply by berendkleinhaneveld** [reply]
Actually, this https://github.com/fork-tongue/collagraph/blob/master/collagraph/renderers/pyside/objects/itemmodel.py would be a better example. After the weekend I can help out with this. Should be fairly straightforward to implement.

**Reply by jamesdbrock** [reply]
Is there a way to do this without making changes to https://github.com/fork-tongue/collagraph/blob/ba447e1e5b1ac5d31fc44e732d16bd28275e0f47/collagraph/renderers/pyside_renderer.py#L47 ?

**Reply by berendkleinhaneveld** [reply]
No, that is exactly where the new methods should be registered. We could maybe standardize a bit on the interface and then load all methods from the modules within the `objects` folder, such that it would pick up new methods from new modules automatically.

**Reply by jamesdbrock** [reply]
Yeah it would be nice if new custom `QWidget` registrations could be done in user code. __re-wx__ has a way to register type-level dispatch of custom widgets from user code with decorators. https://github.com/chriskiehl/re-wx/blob/main/rewx/dispatch.py

In user code it looks like this. https://github.com/chriskiehl/re-wx/issues/21
